TestRoutingKey is flaky

It fails on driver matrix from time to time
`cassandra_test.go:2321: failed to get routing key info due to error: no metadata available
09:01:38  --- FAIL: TestRoutingKey (1.23s)`

https://jenkins.scylladb.com/job/scylla-2025.4/job/driver-tests/job/gocql-driver-matrix-test/13/

--- a/cassandra_test.go
+++ b/cassandra_test.go
@@ -2661,16 +2661,19 @@ func TestRoutingKey(t *testing.T) {
 	session := createSession(t)
 	defer session.Close()
 
-	if err := createTable(session, "CREATE TABLE gocql_test.test_single_routing_key (first_id int, second_id int, PRIMARY KEY (first_id, second_id))"); err != nil {
+	singleTable := testTableName(t, "single")
+	compositeTable := testTableName(t, "composite")
+
+	if err := createTable(session, fmt.Sprintf("CREATE TABLE gocql_test.%s (first_id int, second_id int, PRIMARY KEY (first_id, second_id))", singleTable)); err != nil {
 		t.Fatalf("failed to create table with error '%v'", err)
 	}
-	if err := createTable(session, "CREATE TABLE gocql_test.test_composite_routing_key (first_id int, second_id int, PRIMARY KEY ((first_id, second_id)))"); err != nil {
+	if err := createTable(session, fmt.Sprintf("CREATE TABLE gocql_test.%s (first_id int, second_id int, PRIMARY KEY ((first_id, second_id)))", compositeTable)); err != nil {
 		t.Fatalf("failed to create table with error '%v'", err)
 	}
 
 	initCacheSize := session.routingKeyInfoCache.lru.Len()
 
-	routingKeyInfo, err := session.routingKeyInfo(context.Background(), "SELECT * FROM test_single_routing_key WHERE second_id=? AND first_id=?", time.Second)
+	routingKeyInfo, err := session.routingKeyInfo(context.Background(), fmt.Sprintf("SELECT * FROM %s WHERE second_id=? AND first_id=?", singleTable), time.Second)
 	if err != nil {
 		t.Fatalf("failed to get routing key info due to error: %v", err)
 	}
@@ -2696,7 +2699,7 @@ func TestRoutingKey(t *testing.T) {
 	// verify the cache is working
 	routingKeyInfo, err = session.routingKeyInfo(
 		context.Background(),
-		"SELECT * FROM test_single_routing_key WHERE second_id=? AND first_id=?",
+		fmt.Sprintf("SELECT * FROM %s WHERE second_id=? AND first_id=?", singleTable),
 		// Routing info will be pulled from cached prepared statement, it should work with minimal timeout
 		time.Nanosecond)
 	if err != nil {
@@ -2722,7 +2725,7 @@ func TestRoutingKey(t *testing.T) {
 		t.Errorf("Expected cache size to be %d but was %d", initCacheSize+1, cacheSize)
 	}
 
-	query := session.Query("SELECT * FROM test_single_routing_key WHERE second_id=? AND first_id=?", 1, 2)
+	query := session.Query(fmt.Sprintf("SELECT * FROM %s WHERE second_id=? AND first_id=?", singleTable), 1, 2)
 	routingKey, err := query.GetRoutingKey()
 	if err != nil {
 		t.Fatalf("Failed to get routing key due to error: %v", err)
@@ -2734,7 +2737,7 @@ func TestRoutingKey(t *testing.T) {
 
 	routingKeyInfo, err = session.routingKeyInfo(
 		context.Background(),
-		"SELECT * FROM test_composite_routing_key WHERE second_id=? AND first_id=?",
+		fmt.Sprintf("SELECT * FROM %s WHERE second_id=? AND first_id=?", compositeTable),
 		time.Second)
 	if err != nil {
 		t.Fatalf("failed to get routing key info due to error: %v", err)
@@ -2767,7 +2770,7 @@ func TestRoutingKey(t *testing.T) {
 		t.Fatalf("Expected routing key types[0] to be %v but was %v", TypeInt, routingKeyInfo.types[1].Type())
 	}
 
-	query = session.Query("SELECT * FROM test_composite_routing_key WHERE second_id=? AND first_id=?", 1, 2)
+	query = session.Query(fmt.Sprintf("SELECT * FROM %s WHERE second_id=? AND first_id=?", compositeTable), 1, 2)
 	routingKey, err = query.GetRoutingKey()
 	if err != nil {
 		t.Fatalf("Failed to get routing key due to error: %v", err)

--- a/common_test.go
+++ b/common_test.go
@@ -198,9 +198,12 @@ func createTable(s *Session, table string) error {
 		return err
 	}
 
-	// Invalidate the keyspace schema cache so that subsequent metadata lookups
-	// reload from the cluster, avoiding races with the debounced schema event.
-	if ks := extractKeyspaceFromDDL(table); ks != "" {
+	// Invalidate schema cache to avoid races with debounced schema events.
+	ks := extractKeyspaceFromDDL(table)
+	if ks == "" {
+		ks = s.cfg.Keyspace
+	}
+	if ks != "" {
 		s.metadataDescriber.invalidateKeyspaceSchema(ks)
 	}
 
@@ -216,8 +219,7 @@ func extractKeyspaceFromDDL(ddl string) string {
 		return ""
 	}
 	rest := strings.TrimSpace(ddl[idx+len("TABLE"):])
-	// DDL may contain optional "IF NOT EXISTS" (CREATE) or "IF EXISTS" (DROP)
-	// between the TABLE keyword and the table name — skip past them.
+	// Skip optional "IF [NOT] EXISTS" between TABLE and the name.
 	upperRest := strings.ToUpper(rest)
 	if strings.HasPrefix(upperRest, "IF NOT EXISTS") {
 		rest = strings.TrimSpace(rest[len("IF NOT EXISTS"):])
@@ -442,6 +444,35 @@ func createAggregate(t *testing.T, session *Session) {
 	`).Exec(); err != nil {
 		t.Fatalf("failed to create aggregate with err: %v", err)
 	}
+}
+
+const maxCQLIdentifierLen = 48
+
+// testTableName builds a CQL-safe table name from t.Name() and optional parts.
+// Truncates to 48 chars (CQL limit) using <first20>_<last20> when needed.
+func testTableName(t testing.TB, parts ...string) string {
+	name := strings.ToLower(t.Name())
+	for _, p := range parts {
+		name += "_" + strings.ToLower(p)
+	}
+
+	var b strings.Builder
+	prevUnderscore := false
+	for _, r := range name {
+		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
+			b.WriteRune(r)
+			prevUnderscore = false
+		} else if !prevUnderscore {
+			b.WriteByte('_')
+			prevUnderscore = true
+		}
+	}
+	name = strings.Trim(b.String(), "_")
+
+	if len(name) > maxCQLIdentifierLen {
+		name = name[:20] + "_" + name[len(name)-20:]
+	}
+	return name
 }
 
 func staticAddressTranslator(newAddr net.IP, newPort int) AddressTranslator {

--- a/scylla_cdc.go
+++ b/scylla_cdc.go
@@ -82,10 +82,7 @@ func scyllaIsCdcTable(session *Session, keyspaceName, tableName string) (bool, e
 		return false, nil
 	}
 
-	// Get the table metadata to see if it has the cdc partitioner set.
-	// Use TableMetadata (which routes through GetTable) instead of
-	// GetKeyspace + Tables[name] to properly handle cache invalidation
-	// when a table was recently created or altered.
+	// Check if the table has the CDC partitioner set.
 	tableMeta, err := session.TableMetadata(keyspaceName, tableName)
 	if err != nil {
 		return false, err

--- a/session.go
+++ b/session.go
@@ -817,6 +817,14 @@ func (s *Session) routingKeyInfo(ctx context.Context, stmt string, requestTimeou
 	table := info.request.table
 	keyspace := info.request.keyspace
 
+	// Fall back to per-column metadata when FlagGlobalTableSpec is not set.
+	if keyspace == "" && len(info.request.columns) > 0 {
+		keyspace = info.request.columns[0].Keyspace
+	}
+	if table == "" && len(info.request.columns) > 0 {
+		table = info.request.columns[0].Table
+	}
+
 	partitioner, err := scyllaGetTablePartitioner(s, keyspace, table)
 	if err != nil {
 		// don't cache this error
@@ -844,11 +852,7 @@ func (s *Session) routingKeyInfo(ctx context.Context, stmt string, requestTimeou
 		return routingKeyInfo, nil
 	}
 
-	// get the table metadata
-	// Use TableMetadata (which routes through GetTable) instead of
-	// GetKeyspace + Tables[name] to properly handle cache invalidation
-	// when a table was recently created or altered.
-
+	// get the table metadata (uses TableMetadata to handle cache invalidation)
 	var tableMetadata *TableMetadata
 	tableMetadata, inflight.err = s.TableMetadata(keyspace, table)
 	if inflight.err != nil {

--- a/session_unit_test.go
+++ b/session_unit_test.go
@@ -29,6 +29,7 @@ package gocql
 
 import (
 	"context"
+	"strings"
 	"testing"
 )
 
@@ -146,10 +147,8 @@ func TestExtractKeyspaceFromDDL(t *testing.T) {
 	}
 }
 
-// TestTableMetadataAfterInvalidation verifies that Session.TableMetadata
-// (used by routingKeyInfo and scyllaIsCdcTable) properly refreshes table
-// metadata after schema invalidation, unlike the old GetKeyspace + Tables[name]
-// pattern which returned stale data.
+// TestTableMetadataAfterInvalidation verifies that TableMetadata refreshes
+// after schema invalidation.
 func TestTableMetadataAfterInvalidation(t *testing.T) {
 	t.Parallel()
 
@@ -193,8 +192,7 @@ func TestTableMetadataAfterInvalidation(t *testing.T) {
 }
 
 // TestTableMetadataAfterKeyspaceInvalidation verifies that TableMetadata
-// works correctly when the entire keyspace cache is cleared (as done by
-// createTable's new cache invalidation logic).
+// works after the entire keyspace cache is cleared.
 func TestTableMetadataAfterKeyspaceInvalidation(t *testing.T) {
 	t.Parallel()
 
@@ -243,8 +241,7 @@ func newTestSessionForTableMetadata(ctrl *schemaDataMock) *Session {
 }
 
 // TestScyllaIsCdcTableAfterInvalidation verifies that scyllaIsCdcTable
-// properly handles invalidated table metadata (the same bug pattern as
-// routingKeyInfo).
+// handles invalidated table metadata.
 func TestScyllaIsCdcTableAfterInvalidation(t *testing.T) {
 	t.Parallel()
 
@@ -299,5 +296,82 @@ func TestScyllaIsCdcTableNotCdcSuffix(t *testing.T) {
 	}
 	if isCdc {
 		t.Fatal("expected regular_table to not be a CDC table")
+	}
+}
+
+func TestTestTableName(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name  string
+		parts []string
+		want  string
+	}{
+		{
+			name: "basic",
+			want: "testtesttablename_basic",
+		},
+		{
+			name:  "with_parts",
+			parts: []string{"single"},
+			want:  "testtesttablename_with_parts_single",
+		},
+		{
+			name:  "multiple_parts",
+			parts: []string{"foo", "bar"},
+			want:  "testtesttablename_multiple_parts_foo_bar",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := testTableName(t, tt.parts...)
+			if got != tt.want {
+				t.Errorf("testTableName() = %q, want %q", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTestTableNameSanitizesSpecialChars(t *testing.T) {
+	t.Parallel()
+
+	// t.Name() for subtests contains '/', verify it's sanitized.
+	t.Run("sub/with/slashes", func(t *testing.T) {
+		got := testTableName(t)
+		if strings.Contains(got, "/") {
+			t.Errorf("expected no slashes, got %q", got)
+		}
+		// Consecutive separators from / should be collapsed.
+		if strings.Contains(got, "__") {
+			t.Errorf("expected no consecutive underscores, got %q", got)
+		}
+	})
+}
+
+func TestTestTableNameTruncation(t *testing.T) {
+	t.Parallel()
+
+	// Build a subtest name that forces the result over 48 chars.
+	long := "abcdefghijklmnopqrstuvwxyz0123456789abcdefghijklmnopqrstuvwxyz"
+	t.Run(long, func(t *testing.T) {
+		got := testTableName(t, "extra")
+		if len(got) > maxCQLIdentifierLen {
+			t.Errorf("len = %d, want <= %d; value = %q", len(got), maxCQLIdentifierLen, got)
+		}
+		// Should contain chars from both the start and end.
+		if got[:5] != "testt" {
+			t.Errorf("expected prefix from test name, got %q", got)
+		}
+	})
+}
+
+func TestTestTableNameUniqueness(t *testing.T) {
+	t.Parallel()
+
+	a := testTableName(t, "alpha")
+	b := testTableName(t, "beta")
+	if a == b {
+		t.Errorf("expected different names, both got %q", a)
 	}
 }

